### PR TITLE
Add basic web client configuration

### PR DESCRIPTION
**Pull request by dkanada**

### CODE REVIEW

File: src/scripts/settings/webSettings.js [hunk 1]
```diff
@@ -0,0 +1,30 @@
+define(['appStorage', 'events'], function (appStorage, events) {
+    'use strict';
+
+    function readConfig(path, callback) {
+        var file = new XMLHttpRequest();
```
Comment (thornbill) [inline src/scripts/settings/webSettings.js]: Should use fetch and promises instead of callbacks.

File: src/components/skinManager.js [hunk 2]
```diff
@@ -1,4 +1,4 @@
-define(['apphost', 'userSettings', 'browser', 'events', 'pluginManager', 'backdrop', 'globalize', 'require', 'appSettings'], function (appHost, userSettings, browser, events, pluginManager, backdrop, globalize, require, appSettings) {
+define(['apphost', 'userSettings', 'browser', 'events', 'pluginManager', 'backdrop', 'globalize', 'require', 'appSettings', 'webSettings'], function (appHost, userSettings, browser, events, pluginManager, backdrop, globalize, require, appSettings, webSettings) {
```
Comment (JustAMan) [inline src/components/skinManager.js]: why add?

File: src/scripts/settings/webSettings.js [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+
+    var data = {};
+
+    function WebSettings() {
+        fetch("/config.json").then(function (response) {
+            data = response.json();
+        })
+    }
+
+    WebSettings.prototype.getMultiServer = function () {
```
Comment (JustAMan) [inline src/scripts/settings/webSettings.js]:
```suggestion
    WebSettings.prototype.allowMultiServer = function () {
```

File: src/strings/en-us.json [hunk 4]
```diff
@@ -15,7 +15,7 @@
     "Alerts": "Alerts",
     "All": "All",
     "AllChannels": "All channels",
-    "AllComplexFormats": "All complex formats (ASS, SSA, VOBSUB, PGS, SUB/IDX, etc.)",
+    "AllComplexFormats": "All Complex Formats (ASS, SSA, VOBSUB, PGS, SUB, IDX)",
```
Comment (JustAMan) [inline src/strings/en-us.json]: why do you edit language files when adding configuration? don't mix unrelated stuff please :)

File: src/scripts/settings/webSettings.js (hunk 3 above)
Comment (dkanada) [reply src/scripts/settings/webSettings.js]: I figured we could use get here because we will never add a set method to these options.

File: src/strings/en-us.json (hunk 4 above)
Comment (dkanada) [reply src/strings/en-us.json]: I made these changes because I was originally adding some configuration options to the subtitles options page, I wouldn't normally do that.

File: src/components/skinManager.js (hunk 2 above)
Comment (dkanada) [reply src/components/skinManager.js]: good catch

File: src/scripts/settings/webSettings.js (hunk 3 above)
Comment (JustAMan) [reply src/scripts/settings/webSettings.js]: I still don't like the verb `get` here, because it looks a bit confusing to me what it "gets".
Maybe rename to `isMultiServerEnabled` to be extra-clear?

File: src/strings/en-us.json (hunk 4 above)
Comment (JustAMan) [reply src/strings/en-us.json]: I'll believe you... this time 👅

File: src/scripts/settings/webSettings.js [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+    var data = {};
+
+    function WebSettings() {
+        fetch("/config.json").then(function (response) {
+            data = response.json();
+        })
+    }
+
+    WebSettings.prototype.enableMultiServer = function () {
+        return data.multiServer || false;
```
Comment (thornbill) [inline src/scripts/settings/webSettings.js]: Since getting the config is async this could randomly return the default if the fetch has not completed when this function is called which doesn’t seem ideal. I wonder if it would be better to have a single function `getConfig()` or something that returns a `Promise` that will resolve with the complete config object.

Comment (thornbill) [inline src/scripts/settings/webSettings.js]: Since getting the config is async this could randomly return the default if the fetch has not completed when this function is called which doesn’t seem ideal. I wonder if it would be better to have a single function `getConfig()` or something that returns a `Promise` that will resolve with the complete config object.

File: src/scripts/settings/webSettings.js [hunk 6]
```diff
@@ -1,16 +1,26 @@
...
+        if (data) {
+            return data;
+        }
 
-    function WebSettings() {
         fetch("/config.json").then(function (response) {
             data = response.json();
         })
+
+        return data;
```
Comment (JustAMan) [inline src/scripts/settings/webSettings.js]: this feels wrong... if anything your `getConfig` should be a promise or something...

File: src/components/displaysettings/displaysettings.template.html [hunk 7]
```diff
@@ -63,7 +63,7 @@ <h2 class="sectionTitle">
 
     <div class="selectContainer fldDateTimeLocale hide">
         <select is="emby-select" class="selectDateTimeLocale" label="${LabelDateTimeLocale}">
-            <option value="">${AutoBasedOnLanguageSetting}</option>
+            <option value="">${Auto}</option>
```
Comment (MrTimscampi) [inline src/components/displaysettings/displaysettings.template.html:66]: I feel like it's clearer to let it be "Auto (based on language setting)", though I'd remove the "Auto" part.

File: src/scripts/settings/appSettings.js [hunk 8]
```diff
@@ -20,6 +20,14 @@ define(['appStorage', 'events'], function (appStorage, events) {
         return this.get('enableAutoLogin') !== 'false';
     };
 
+    AppSettings.prototype.enableSystemExternalPlayers = function (val) {
+        if (val != null) {
```
Comment (MrTimscampi) [inline src/scripts/settings/appSettings.js:24]:
```suggestion
        if (val !== null) {
```

A strict comparison is usually better.

**Comment by dmitrylyzo** [comment]
@dkanada why did you refuse config caching (`data`)? You could simply return `Promise.resolve(data)` if `data` is fetched. At current moment, if there will be more `getConfig` calls, it will be fetched each time.

**Comment by dkanada** [comment]
The intent is to apply config changes immediately, which won't work if any kind of browser cache is used during the fetch.

File: src/scripts/settings/appSettings.js (hunk 8 above)
Comment (dkanada) [reply src/scripts/settings/appSettings.js:24]: good catch, missed that

File: src/components/displaysettings/displaysettings.template.html (hunk 7 above)
Comment (dkanada) [reply src/components/displaysettings/displaysettings.template.html:66]: I was mainly trying to reduce the amount of strings I think, because `auto` already implies automatic detection using some metric and we don't explain any similar instances.

File: src/scripts/settings/webSettings.js [hunk 9]
```diff
@@ -1,27 +1,11 @@
...
-    }
-
-    WebSettings.prototype.enableMultiServer = function () {
-        return getConfig().multiServer || false;
-    };
-
-    return new WebSettings();
-});
+function getConfig() {
+    return fetch("/config.json?nocache=" + new Date().getUTCMilliseconds()).then(function (response) {
```
Comment (JustAMan) [inline src/scripts/settings/webSettings.js:2]: Any reason for not storing the value of response to a local variable?
IMHO it should request settings once and use them until a page is completely refreshed.

Comment (dkanada) [reply src/scripts/settings/webSettings.js:2]: > The intent is to apply config changes immediately, which won't work if any kind of browser cache is used during the fetch.

Comment (JustAMan) [reply src/scripts/settings/webSettings.js:2]: I see the intent, but it would make running the UI slower by adding one more request.
The case when you update the settings is usually rare, you don't go changing them all the time.

So refreshing a page manually when settings are changed isn't a big deal, but adding more lag is IMHO.

Comment (dkanada) [reply src/scripts/settings/webSettings.js:2]: Refreshing the page manually will always hit the browser cache with the same URL and we can't ask every user to update their browser cache whenever they change these settings.

Comment (dmitrylyzo) [reply src/scripts/settings/webSettings.js:2]: > You could simply return Promise.resolve(data) if data is fetched.

```js
let data;

function getConfig() {
    if (data) {
        return Promise.resolve(data);
    }

    return fetch("/config.json?nocache=" + new Date().getUTCMilliseconds()).then(function (response) {
        data = response.json();
        return data;
    });
}
...
```

Comment (dkanada) [reply src/scripts/settings/webSettings.js:2]: That would be a nice middle ground, but it depends on if we want to refresh the settings during a user session or not. This wouldn't refresh until they fully reloaded the website.

Comment (dkanada) [reply src/scripts/settings/webSettings.js:2]: I went with this method, I think once per session should be enough now that you mention it.
